plugins/genesys/fu-genesys-scaler-device.c

```diff
@@ -0,0 +1,1806 @@
+						   error)) {
+			g_prefix_error(error, "error getting public key: ");
+			return FALSE;
+		}
+
+		g_usleep(100000); /* 100ms */
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_genesys_scaler_device_read_flash(FuGenesysScalerDevice *self,
+				    FuProgress *progress,
+				    guint addr,
+				    guint8 *buf,
+				    guint bufsz,
+				    GError **error)
+{
+	FuDevice *parent_device = fu_device_get_parent(FU_DEVICE(self));
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(parent_device));
+	guint8 data1[] = {
+	    GENESYS_SCALER_CMD_DATA_WRITE,
+	    0x00, /* Read Data command */
+	    (addr & 0x00ff0000) >> 16,
+	    (addr & 0x0000ff00) >> 8,
+	    (addr & 0x000000ff),
+	};
+	guint8 data2[] = {
+	    GENESYS_SCALER_CMD_DATA_READ,
+	};
+	guint8 data3[] = {
+	    GENESYS_SCALER_CMD_DATA_END,
+	};
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	if (!fu_cfi_device_get_cmd(self->cfi_device, FU_CFI_DEVICE_CMD_READ_DATA, &data1[1], error))
+		return FALSE;
+
+	if (!g_usb_device_control_transfer(usb_device,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_DEVICE,
+					   self->vc.req_write,
+					   0x0000,	  /* value */
+					   0x0000,	  /* idx */
+					   data1,	  /* data */
+					   sizeof(data1), /* data length */
+					   NULL,	  /* actual length */
+					   GENESYS_SCALER_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "error reading flash at 0x%06x: ", addr);
+		return FALSE;
+	}
+
+	if (!g_usb_device_control_transfer(usb_device,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_DEVICE,
+					   self->vc.req_write,
+					   0x0000,	  /* value */
+					   0x0000,	  /* idx */
+					   data2,	  /* data */
+					   sizeof(data2), /* data length */
+					   NULL,	  /* actual length */
+					   GENESYS_SCALER_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "error reading flash at 0x%06x: ", addr);
+		return FALSE;
+	}
+
+	chunks = fu_chunk_array_mutable_new(buf, bufsz, addr, 0, self->transfer_size);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+
+		if (!g_usb_device_control_transfer(usb_device,
+						   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+						   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+						   G_USB_DEVICE_RECIPIENT_DEVICE,
+						   self->vc.req_read,
+						   0x0000,		       /* value */
+						   0x0000,		       /* idx */
+						   fu_chunk_get_data_out(chk), /* data */
+						   fu_chunk_get_data_sz(chk),  /* data length */
+						   NULL,		       /* actual length */
+						   GENESYS_SCALER_USB_TIMEOUT,
+						   NULL,
+						   error)) {
+			g_prefix_error(error,
+				       "error reading flash at 0x%06x: ",
+				       fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		fu_progress_set_percentage_full(progress,
+						i * self->transfer_size + fu_chunk_get_data_sz(chk),
+						bufsz);
```

Comment:
can you use `fu_progress_set_steps(progress, chunks->len);` and then just do `fu_progress_step_done(progress);` here?